python_modules/dagster/dagster/_core/definitions/module_loaders/utils.py

```diff
@@ -37,25 +37,6 @@ def find_subclasses_in_module(
             yield value
 
 
-def key_iterator(
```

Comment:
Given there's currently just a single place where we need to do this, and most of this ugliness gets cleaned up by removing the `include(d)_targeted_keys` param, I think it makes more sense to consolidate this logic here for now

Reply:
This isn't actually the only place where we need to do this, is my take. I needed to do the same thing in Airlift, and anything that operates on an unresolved `Definitions` object will need to do the same thing. 

I think the reality is that in each of the places where we likely encounter it there's a ternary that I didn't break by doing this.

Reply:
seems like this only gets you half of the way there on the annoyance removal though, right? because you still have to itertools.chain() all of your various iterators together into a single iterator or do the dastardly double list comprehension that we currently have

if we already have a lovely `map_asset_specs()` method that handles the iteration and sub-iteration for you on these heterogenous lists, maybe we just supply an `iterate_asset_specs()` method to do the same, just without the mapping bit?

that seems a bit more general-purpose to me, and it'd let you do:

```python
all_asset_keys = {
    spec.key
    for spec in iterate_asset_specs(self.assets_defs_specs_and_checks_defs)
}
```

which is a lot easier to comprehend.

we can keep it as an internal utility for now so that airlift can make use of it

Reply:
I guess the only confounding factor there is source assets (since we don't support this pattern on cacheables right now anyway). and I guess checks defs would just return nothing, which is maybe fine?